pkg/routeagent_driver/handlers/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,325 @@
+
+func (h *controller) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	if !h.config.HealthCheckerEnabled || h.State().IsOnGateway() {
+		return nil
+	}
+
+	return h.processEndpointCreatedOrUpdated(endpoint)
+}
+
+func (h *controller) RemoteEndpointUpdated(endpoint *submarinerv1.Endpoint) error {
+	if !h.config.HealthCheckerEnabled || h.State().IsOnGateway() {
+		return nil
+	}
+
+	return h.processEndpointCreatedOrUpdated(endpoint)
+}
+
+func (h *controller) processEndpointCreatedOrUpdated(endpoint *submarinerv1.Endpoint) error {
+	logger.Infof("Processing Endpoint: %#v", endpoint)
+
+	if endpoint.Spec.HealthCheckIP == "" || endpoint.Spec.CableName == "" {
+		logger.Infof("HealthCheckIP (%q) and/or CableName (%q) for Endpoint %q empty - will not monitor endpoint health",
+			endpoint.Spec.HealthCheckIP, endpoint.Spec.CableName, endpoint.Name)
+		return nil
+	}
+
+	h.Lock()
+	defer h.Unlock()
+
+	if pingerObject, found := h.pingers[endpoint.Spec.CableName]; found {
+		if pingerObject.GetIP() == endpoint.Spec.HealthCheckIP {
+			return nil
+		}
+
+		logger.Infof("HealthChecker is already running for %q - stopping", endpoint.Name)
+		pingerObject.Stop()
+		delete(h.pingers, endpoint.Spec.CableName)
+	}
+
+	pingerConfig := pinger.Config{
+		IP:                 endpoint.Spec.HealthCheckIP,
+		MaxPacketLossCount: h.config.MaxPacketLossCount,
+	}
+
+	if h.config.PingInterval != 0 {
+		pingerConfig.Interval = time.Second * time.Duration(h.config.PingInterval)
+	}
+
+	if h.config.MaxPacketLossCount != 0 {
+		pingerConfig.MaxPacketLossCount = h.config.MaxPacketLossCount
+	}
+
+	newPingerFunc := h.config.NewPinger
+	if newPingerFunc == nil {
+		newPingerFunc = pinger.NewPinger
+	}
+
+	pingerObject := newPingerFunc(pingerConfig)
+	h.pingers[endpoint.Spec.CableName] = pingerObject
+	pingerObject.Start()
+
+	logger.Infof("HealthChecker started pinger for CableName: %q with HealthCheckIP %q",
+		endpoint.Spec.CableName, endpoint.Spec.HealthCheckIP)
+
+	return nil
+}
+
+func (h *controller) RemoteEndpointDeleted(endpoint *submarinerv1.Endpoint) error {
+	h.Lock()
+	defer h.Unlock()
+
+	if pingerObject, found := h.pingers[endpoint.Spec.CableName]; found {
+		pingerObject.Stop()
+		delete(h.pingers, endpoint.Spec.CableName)
+	}
+
+	return nil
+}
+
+func (h *controller) Init() error {
+	go func() {
+		wait.Until(func() {
+			h.Lock()
+			defer h.Unlock()
+
+			h.syncRouteAgentStatus()
+		}, RouteAgentUpdateInterval, h.stopCh)
+	}()
+
+	return nil
+}
+
+// TransitionToNonGateway is called once for each transition of the local node from Gateway to a non-Gateway.
+func (h *controller) TransitionToNonGateway() error {
+	h.Lock()
+	defer h.Unlock()
+
+	if h.config.HealthCheckerEnabled {
+		remoteEndpoints := h.State().GetRemoteEndpoints()
+
+		for i := range remoteEndpoints {
+			err := h.processEndpointCreatedOrUpdated(&remoteEndpoints[i])
+			if err != nil {
+				logger.Warningf("Error processing remote endpoint %s: %v", remoteEndpoints[i].Name, err)
```

Comment:
I think we should return the error here.

Reply:
Done.